### DISCUSSION TITLE
Commissioner Session ID TLV and Commissioner IPv6 address do not match.

### DISCUSSION THREAD

**Discussion by y-morgunov** [openthread/openthread#8245] [Q&A] [answered]
Good afternoon!

I have a misunderstanding of how I should work with the commissioner.

Situation: when starting the device, the commissioner starts after receiving the OT_CHANGED_THREAD_ROLE event.
However, if the network already has connected devices, the address that the commissioner receives does not match what it should be. According to the specification, the last byte of the address must be equal to (Session ID TLV) % 8.

Here's what it should be:
```
Session id tlv - 0b026024
ip addr: fdb6:1935:91af:450d:0:ff:fe00:fc34
```

But in fact, ip is random.
```
Session id tlv - 0b026024
ip addr: fdb6:1935:91af:450d:0:ff:fe00:fc33
```

Most likely the problem is that netdata is updated later than the commissioner starts.

This is proved by the fact that if I do not start the commissioner on the OT_CHANGED_THREAD_ROLE event, but start it some time after the role change, then the ip is set as expected.

This problem causes several problems, firstly I use the IPv6 of the commissioner to determine who to send udp packets to. 
Secondly, because of this, the joiner cannot be connected to the network. "Failed to process Discovery Request: Security" error occurs

When should I start commissioner? What are your thoughts on this matter?

Thank you!

P.S.  I am using nrf52840 on PCA10056

**Comment by jwhui** [comment]
@y-morgunov , thanks for raising this issue. It's yet clear why this behavior is occurring. Can you help share a sniffer packet capture so that we can better analyze the scenario?

**Reply by y-morgunov** [reply]
@jwhui wireshark - https://drive.google.com/file/d/1cjVwnebvsX9jl_rHH4vktmA-dXBB6cCf/view?usp=sharing
Network key - 6dc9709b0a9a4f548803d78645cfa089

I forgot to say I am using nrf52840 on PCA10056

Timings:

- Commissioner board start at 0s;
- Enabling 3 nodes ~19s;
  ipaddr - fdc6:b9cb:4722:fec8:0:ff:fe00:fc32
  netdata - 080b090220000b02f0c2080100
- Commissioner board reset at ~76s;
  ipaddr - fdc6:b9cb:4722:fec8:0:ff:fe00:fc30
  netdata - 080b090220000b02f0c2080100
- Joining start at ~125s

**Comment by jwhui** [comment, marked as answer]
This issue occurs when the Leader is restoring network data after a restart. After a restart, the Leader will restore network data by retrieving it from a neighboring node. If the Commissioner registers itself before it properly restores network data, the Commissioner information in the network data will be overwritten.

One way to fix this issue is by having the Leader reject updates to the network data until after it has properly restored it. I submitted #8332 to implement this.
